Add prometheus metrics

mozzieongit opened the pull request:
This adds support for Prometheus metrics, exposed using `libevent`'s HTTP server
implementation. It is feature gated with `--enable-prometheus-metrics`.

The statistics export and socket code is based on code from `remote.c`.

Here is an excerpt from the exported metrics:
```
nsd_queries_total{server="0"} 27
nsd_queries_total{server="1"} 22
nsd_queries_by_type_total{type="A"} 42
nsd_queries_by_class_total{class="IN"} 49
nsd_queries_by_opcode_total{opcode="QUERY"} 49
nsd_queries_by_rcode_total{rcode="NOERROR"} 48
nsd_queries_by_transport_total{transport="udp"} 25
nsd_queries_with_edns_total 49
nsd_queries_with_edns_failed_total 0

nsd_connections_total{transport="tcp"} 24
nsd_connections_total{transport="tls6"} 0

nsd_xfr_requests_served_total{xfrtype="AXFR"} 0
nsd_xfr_requests_served_total{xfrtype="IXFR"} 0

nsd_queries_dropped_total 0
nsd_queries_rx_failed_total 0

nsd_answers_tx_failed_total 0
nsd_answers_without_aa_total 0
nsd_answers_truncated_total 0

nsd_time_up_seconds_total 970.392208

nsd_size_db_on_disk_bytes 0
nsd_size_db_in_mem_bytes 0
nsd_size_xfrd_in_mem_bytes 1163464
nsd_size_config_on_disk_bytes 0
nsd_size_config_in_mem_bytes 2208

nsd_zones_primary 1
nsd_zones_secondary 0

nsd_zonestats_examplezone_queries_total 48
nsd_zonestats_examplezone_queries_by_type_total{type="A"} 41
```

mozzieongit committed:
Add metrics configure option and make target

mozzieongit committed:
Add metrics configuration options

mozzieongit committed:
Enable use of metrics endpoint

mozzieongit committed:
Implement prometheus metrics endpoint

mozzieongit committed:
Update metrics config documentation

mozzieongit committed:
Rename metrics and add help texts

Metrics should not contain duplicate data, summing all metrics of
a given name (regardless of labels) needs to be meaningful. Therefore
the counter metrics based on different DNS aspects have been split into
separate metric names.

mozzieongit committed:
Deduplicate stats code by using a prefix string

mozzieongit commented:
>     1. The elapsed time (`nsd_time_elapsed_seconds`) doesn't have much value
>        currently, and doesn't reset when `nsd-control stats` is used, so I would
>        probably remove it. However, it could be beneficial to reset this time
>        whenever `nsd-control` resets the stats.
I disabled the elapsed time metric, until I point it to the remote-control time value.
 
>     2. For the zonestats block: Should the label name "zone" maybe renamed to
>        "zonegroup" or something else?
I used `zonestats`, to copy the name from the config file. 

>     3. Currently, the metrics don't have any HELP text (see [Prometheus Exposition Formats](https://prometheus.io/docs/instrumenting/exposition_formats/#comments-help-text-and-type-information))
Help and type annotations added.

mozzieongit committed:
Update comments and remove unused function

mozzieongit committed:
Rename metrics feature flag to prometheus-metrics

wcawijngaards commented on options.c:
```diff
@@ -162,6 +162,12 @@ nsd_options_create(region_type* region)
 	opt->server_cert_file = CONFIGDIR"/nsd_server.pem";
 	opt->control_key_file = CONFIGDIR"/nsd_control.key";
 	opt->control_cert_file = CONFIGDIR"/nsd_control.pem";
+#ifdef USE_METRICS
```
The metrics options are not in nsd-checkconf.c, in the functions config_print_zone and config_test_print_server the options also need to be handled.

wcawijngaards commented on metrics.c:
```diff
@@ -0,0 +1,815 @@
+	                           "Uptime since server boot in seconds.", "counter");
+	evbuffer_add_printf(buf, "nsd_time_up_seconds_total %lu.%6.6lu\n",
+		(unsigned long)uptime.tv_sec, (unsigned long)uptime.tv_usec);
+	/* TODO: re-add when elapsed time resetting on nsd-control stats is implemented
```
Can this have for example an implementation with `if(clear) xfrd->nsd->rc->stats_time = *now;`, similar to remote.c:3230. The comment could also be about something else.

wcawijngaards commented on metrics.c:
```diff
@@ -0,0 +1,815 @@
+
+/** process the statistics and write them into the buffer */
+static void
+process_stats(struct evbuffer *buf, xfrd_state_type* xfrd, int peek)
```
The routines are copies of code at remote.c, I do not think that copying code that manages the inter-process communication part of the statistics like that is a very good idea. It would be better to make the routine public in remote.c and add its definition to remote.h and then call it here. The process_stats() function then needs to be changed to allow the different print_stats call than the print_stats call from remote.c.

wcawijngaards commented on nsd.conf.5.in:
```diff
@@ -603,6 +603,30 @@ The port number for proxy protocol service. If the statement is given multiple
+The port number for the HTTP service. Default is 9100.
+.TP
+.B metrics\-path:\fR <string>
+The HTTP path to expose the metrics at. Default is /metrics.
```
Perhaps the path name extent is more precise with a quoted string.

```suggestion
The HTTP path to expose the metrics at. Default is "/metrics".
```

wcawijngaards reviewed (changes requested):
The code looks good, and in general I like it, but I would want to change the copy of the statistics management code from remote.c into metrics.c, instead expose one or more static routines from remote.c and call them from metrics.c, making the code more understandable and maintainable.

mozzieongit committed:
Put metrics URL path in quotes

mozzieongit committed:
Expose nsd stats allocation in remote and use from metrics

mozzieongit committed:
Combine metrics and remote process_stats function

mozzieongit committed:
Add metric for time since last stats reset by nsd-control

mozzieongit committed:
Expose opcode2str function for use from metrics code

mozzieongit committed:
Expose timeval_subtract

mozzieongit committed:
Deduplicate zonestats collection

mozzieongit committed:
Add metrics options to nsd-checkconf

mozzieongit committed:
Fix compilation with metrics disabled

mozzieongit committed:
Fix warnings from checking enabled

mozzieongit commented on remote.c:
```diff
@@ -3220,7 +3221,17 @@ process_stats(RES* ssl, xfrd_state_type* xfrd, int peek)
+#ifdef USE_METRICS
+	if (evbuf) {
+		metrics_print_stats(evbuf, xfrd, &stattime, !peek, &total, zonestats,
+		                    &xfrd->nsd->rc->stats_time);
```
Found an issue:
- one of `ssl` and `evbuffer` must be NULL, or,
- if both are non-NULL, `peek` needs to be set to `1` for the first invocation of the two `...print_stats`, so that the stats get cleared after both stats prints are done.

I think it makes sense to put an assert there to make sure only one of the two is set.

wcawijngaards reviewed (approved):
The fixes look good, I like how the deduplication of the code turned out! An added assert would be fine.

mozzieongit committed:
Assert that only one stats output is used

mozzieongit committed:
Add test for prometheus metrics with remote-control enabled

mozzieongit committed:
Add test for prometheus metrics with remote-control disabled

I'm adding a separate test for this because the metrics endpoint would
print a time from the remote control endpoint, which would be
uninitialized when remote-control is disabled and should therefore be
omitted.

mozzieongit committed:
Use metrics stats_time for elapsed time when remote-control is disabled

mozzieongit commented:
I added the tests for the metrics. One with remote-control enabled, resetting the stats, and basically just sanity checking that the metrics also reset. One with rc disabled just to make sure it works in both cases.
While writing the test, I ran into a segfault when rc was disabled, therefore the fix in f08dec9.

wcawijngaards reviewed (approved):
The tests look good! Nice to have tests and bug fixes for the new feature.
